// ng-packagr/lib/ngc/compile-source-files.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const ng = require("@angular/compiler-cli");
const log = require("../util/log");
const create_emit_callback_1 = require("./create-emit-callback");
const redirect_write_file_compiler_host_1 = require("../ts/redirect-write-file-compiler-host");
const cache_compiler_host_1 = require("../ts/cache-compiler-host");
function compileSourceFiles(graph, entryPoint, tsConfig, moduleResolutionCache, stylesheetProcessor, extraOptions, declarationDir) {
    return __awaiter(this, void 0, void 0, function* () {
        log.debug(`ngc (v${ng.VERSION.full})`);
        const tsConfigOptions = Object.assign({}, tsConfig.options, extraOptions);
        let tsCompilerHost = cache_compiler_host_1.cacheCompilerHost(graph, entryPoint, tsConfigOptions, moduleResolutionCache, stylesheetProcessor);
        if (declarationDir) {
            tsCompilerHost = redirect_write_file_compiler_host_1.redirectWriteFileCompilerHost(tsCompilerHost, tsConfigOptions.basePath, declarationDir);
        }
        // ng.CompilerHost
        const ngCompilerHost = ng.createCompilerHost({
            options: tsConfigOptions,
            tsHost: tsCompilerHost
        });
        const emitFlags = tsConfigOptions.declaration ? tsConfig.emitFlags : ng.EmitFlags.JS;
        const scriptTarget = tsConfigOptions.target;
        const cache = entryPoint.cache;
        const oldProgram = cache.oldPrograms && cache.oldPrograms[scriptTarget];
        const ngProgram = ng.createProgram({
            rootNames: tsConfig.rootNames,
            options: tsConfigOptions,
            host: ngCompilerHost,
            oldProgram
        });
        yield ngProgram.loadNgStructureAsync();
        const diagnostics = [
            ...ngProgram.getNgOptionDiagnostics(),
            ...ngProgram.getTsSyntacticDiagnostics(),
            ...ngProgram.getTsSemanticDiagnostics(),
            ...ngProgram.getNgSemanticDiagnostics(),
            ...ngProgram.getNgStructuralDiagnostics()
        ];
        log.debug(`ngc program structure is reused: ${oldProgram ? oldProgram.getTsProgram().structureIsReused : 'No old program'}`);
        cache.oldPrograms = Object.assign({}, cache.oldPrograms, { [scriptTarget]: ngProgram });
        const exitCode = ng.exitCodeFromResult(diagnostics);
        if (exitCode !== 0) {
            throw new Error(ng.formatDiagnostics(diagnostics));
        }
        ngProgram.emit({
            emitCallback: create_emit_callback_1.createEmitCallback(tsConfigOptions),
            emitFlags
        });
    });
}
exports.compileSourceFiles = compileSourceFiles;
